### PR TITLE
Add markdown (.md) highlight support

### PR DESCRIPTION
**Pull request by pmiossec**
## Screenshot

image

### CODE REVIEW

File: Project/Resources/Markdown.xshd [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            </Span>
+            <Span name="Header3" stopateol="true" bold="true" color="DodgerBlue">
+                    <Begin startofline="true">###</Begin>
+            </Span>
+
+            <Span name="Header2" stopateol="true" bold="true" color="Blue">
+                    <Begin startofline="true">##</Begin>
+            </Span>
+
+            <Span name="Header2" stopateol="true" bold="true" color="DarkBlue">
```
Comment (drewnoakes) [inline Project/Resources/Markdown.xshd]:
```suggestion
            <Span name="Header1" stopateol="true" bold="true" color="DarkBlue">
```

File: Project/Resources/Markdown.xshd [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            </Span>
+
+            <Span name="Bold" color="Black" bold="true" stopateol="true">
+                <Begin>**</Begin>
+                <End>**</End>
+            </Span>
+
+            <Span name="Italic" color="Black" italic="true" stopateol="true">
+                <Begin>__</Begin>
+                <End>__</End>
```
Comment (drewnoakes) [inline Project/Resources/Markdown.xshd:79]: Single `_` gives italic too:

```
Apple _banana_ carrot
```

Apple _banana_ carrot

File: Project/Resources/Markdown.xshd [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    <Begin startofline="true">##</Begin>
+            </Span>
+
+            <Span name="Header2" stopateol="true" bold="true" color="DarkBlue">
+                    <Begin startofline="true">#</Begin>
+            </Span>
+
+            <Span name="Bold" color="Black" bold="true" stopateol="true">
+                <Begin>**</Begin>
+                <End>**</End>
```
Comment (drewnoakes) [inline Project/Resources/Markdown.xshd:74]: Single `*` gives italic too:

```
Apple *banana* carrot
```

Apple *banana* carrot

**Comment by drewnoakes** [comment]
Does this support code blocks via ` ``` ` and four-space-indent?

Did you author these? Consider also looking at:

- https://github.com/icsharpcode/AvalonEdit/blob/master/ICSharpCode.AvalonEdit/Highlighting/Resources/MarkDown-Mode.xshd
- https://github.com/ddevault/Halibut/blob/master/Halibut/Highlighting/markdown.xshd

**Comment by pmiossec** [comment]
@drewnoakes 

>Does this support code blocks via ``` and four-space-indent?

No. That's just a basic MarkDown support I made myself. I consider it just good enough but that already add some value.

It was really painful to write as I don't really understand how the parsing works 😢 

>Consider also looking at:

I had a look to see if I found something already working but didn't one. I evaluated 3 or 4 without one working...

> https://github.com/icsharpcode/AvalonEdit/blob/master/ICSharpCode.AvalonEdit/Highlighting/Resources/MarkDown-Mode.xshd

I'm pretty sure I tested this one but it is for a more recent version of the control where the parser have changed and is not compatible. We're stuck with a very old version 😢 

> https://github.com/ddevault/Halibut/blob/master/Halibut/Highlighting/markdown.xshd

I don't remember if I have evaluated this one but I will have a look again (even if the syntax seems not to be the good one)

**Comment by drewnoakes** [comment]
@pmiossec I would not block this PR as it definitely improves support.

> it is for a more recent version of the control where the parser have changed and is not compatible

Oh right, I see. AvalonEdit is the WPF version. It's too bad they're not compatible.

I'm interested in decent markdown support within GE for use in other places too, such as commit message rendering.

**Comment by pmiossec** [comment]
>Consider also looking at:

Both of them don't work, even adapted. I think that's because our current parser doesn't support RegEx

>I'm interested in decent markdown support within GE for use in other places too, such as commit message rendering.

Without regex support, it will be difficult...

>Does this support code blocks via ```

I have added that (and also inline `code`).

> and four-space-indent?

I tried but it didn't worked 😢

**Review by gerhardol — COMMENTED** [review]
OK except for Drew's comments

**Comment by RussKie** [comment]
> AvalonEdit is the WPF version

We can host WPF islands in WinForms (in fact we are already doing this). So if AvalonEdit is something we can use and replace the dated ICSharpCode.TextEditor - sure,  by all means let's discuss.

**Comment by pmiossec** [comment]
I think I've done what I could.

I don't want to touch it anymore. I spent too much time on that frustrating feature.

I know it's far from perfect but good enough for a first try and I don't see how to fix the remaining unsupported features.

I let to someone else the work to improve it (and spend hours on it!) .

**Comment by gerhardol** [comment]
> I think I've done what I could.

Only Drew's three comments....

I believe this is good enough

File: Project/Resources/Markdown.xshd (hunk 2 above)
Comment (pmiossec) [reply Project/Resources/Markdown.xshd:79]: done.

File: Project/Resources/Markdown.xshd (hunk 3 above)
Comment (pmiossec) [reply Project/Resources/Markdown.xshd:74]: Not done because that conflicts with the unordered list items (but added a comment in the file)

**Comment by pmiossec** [comment]
>Only Drew's three comments....

@gerhardol Done. Fixed 2. The last one was not done because I failed to do it...

**Comment by RussKie** [comment]
Drew, are ok with the current state?

File: Project/Resources/Markdown.xshd (hunk 2 above)
Comment (gerhardol) [reply Project/Resources/Markdown.xshd:79]: GH shows two __ still

**Comment by pmiossec** [comment]
I have added a sample file to be able to see the result easily (and for someone who wants to improve the markdown support later)

Doc: What prevent to do a better highlighter:
* no support of regex (from what I've seen)
* can't use space char in string to find because it is an implicit delimiter (and I didn't found a way to work around that)

File: Project/Resources/Markdown.xshd (hunk 2 above)
Comment (pmiossec) [reply Project/Resources/Markdown.xshd:79]: @gerhardol it's comment was to add a new rule to **also** support single '_' that was added elsewhere, so not marking this comment as resolved...

**Review by RussKie — APPROVED** [review]
Awesome! 👍

**Comment by RussKie** [comment]
Merging as is, we can always return to it, if necessary

**Comment by drewnoakes** [comment]
Nice one.

**Pull request merged**